packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,323 @@
+import { assert, fromBase64ToUtf8, hashFile, IsoBuffer, Uint8ArrayToString } from "@fluidframework/common-utils";
+import { ISummaryContext } from "@fluidframework/driver-definitions";
+import { getGitType } from "@fluidframework/protocol-base";
+import * as api from "@fluidframework/protocol-definitions";
+import { PerformanceEvent } from "@fluidframework/telemetry-utils";
+import {
+    IBlob,
+    ISnapshotRequest,
+    ISnapshotResponse,
+    ISnapshotTree,
+    ISnapshotTreeBaseEntry,
+    SnapshotTreeEntry,
+    SnapshotTreeValue,
+    SnapshotType,
+} from "./contracts";
+import { EpochTracker, FetchType } from "./epochTracker";
+import { getUrlAndHeadersWithAuth } from "./getUrlAndHeadersWithAuth";
+import { getWithRetryForTokenRefresh } from "./odspUtils";
+import { TokenFetchOptions } from "./tokenFetch";
+
+/* eslint-disable max-len */
+
+interface IDedupCaches {
+    blobShaToPath: Map<string, string>,
+    pathToBlobSha: Map<string, string>,
+    treesPathToTree: Map<string, api.ISummaryTree>,
+}
+
+export class OdspSummaryUploadManager {
+    // This cache is associated with mapping sha to path for previous summary which belongs to last summary handle.
+    private blobTreeDedupCaches: IDedupCaches = {
+        blobShaToPath: new Map(),
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    private backupBlobTreeDedupCaches: IDedupCaches = {
+        blobShaToPath: new Map(),
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    // Last proposed handle of the uploaded app summary.
+    private lastSummaryProposalHandle: string | undefined;
+
+    constructor(
+        private readonly snapshotUrl: string,
+        private readonly getStorageToken: (options: TokenFetchOptions, name?: string) => Promise<string | null>,
+        private readonly logger: ITelemetryLogger,
+        private readonly epochTracker: EpochTracker,
+        private readonly blobCache: Map<string, IBlob | ArrayBuffer> = new Map(),
+    ) {
+    }
+
+    public async buildCachesForDedup(snapshotTree: api.ISnapshotTree, path: string = ""): Promise<api.ISummaryTree> {
+        assert(Object.keys(snapshotTree.commits).length === 0, "There should not be commit tree entries in snapshot");
+
+        const summaryTree: api.ISummaryTree = {
+            type: api.SummaryType.Tree,
+            tree: {},
+        };
+        for (const [key, value] of Object.entries(snapshotTree.blobs)) {
+            const blobValue = this.blobCache.get(value);
+            assert(blobValue !== undefined, "Blob should exists");
+            const decodedBlobValue = {
+                content: blobValue instanceof ArrayBuffer ? IsoBuffer.from(blobValue).toString("utf8") : fromBase64ToUtf8(blobValue.content),
+                encoding: "utf-8",
+            };
+            const hash = await hashFile(IsoBuffer.from(decodedBlobValue.content, decodedBlobValue.encoding));
```

Comment:
This does not look right. First, Buffer get converted to string and then back to buffer to calculate SHA - that's waste of time.
Second - it's not safe to convert arbitrary binary blob to utf-8 - that will explode, as utf-8 can't be used to represent arbitrary binary data.
We need to switch all the layers (and at some point - IBlob) to using only binary payloads, no strings, no base64, no conversions (other than to utf-8 and back for strings only).

And we should not store blob content in trees that we cache. That's a lot of bytes that should not be needed. If our algorithm relies on actual blobs being there (and recalculating their checksums on every summary), then that's even worse - we spent CPU cycles for nothing. Trees should be shallow. I'm not sure what it means in practice (as we do not have blobID yet for a blob we are about to write). Maybe these trees contains SHAs, i.e. it's a different type of tree we remember, and it has different type of node (what is IBlob today).